Marlin/src/lcd/extensible_ui/ui_api.cpp

```diff
@@ -439,49 +439,89 @@ namespace ExtUI {
   void setTravelAcceleration_mm_s2(const float acc)   { planner.settings.travel_acceleration = acc; }
 
-  #if ENABLED(BABYSTEP_ZPROBE_OFFSET)
-    float getZOffset_mm() {
-      #if ENABLED(BABYSTEP_HOTEND_Z_OFFSET)
-        if (active_extruder != 0)
-          return hotend_offset[Z_AXIS][active_extruder];
-        else
-      #endif
-          return zprobe_zoffset;
+  #if ENABLED(BABYSTEPPING)
+    /**
+     * Converts a mm displacement to a number of steps that is
+     * greater or equal to that displacement (but not less)
+     */
+    inline int16_t nearestSteps(const float mm, const axis_t axis) {
+      const float steps = mm / planner.steps_to_mm[axis];
+      return steps > 0 ? ceil(steps) : floor(steps);
     }
 
-    void setZOffset_mm(const float value) {
-      const float diff = (value - getZOffset_mm()) / planner.steps_to_mm[Z_AXIS];
-      addZOffset_steps(diff > 0 ? ceil(diff) : floor(diff));
-    }
+    /**
+     * This function babysteps an axis during a print.
+     *
+     * When linked_nozzles is false, each nozzle in a multi-nozzle
+     * printer can be babystepped independently of the others. This
+     * lets the user to fine tune the Z-offset and Nozzle Offsets
+     * while observing the first layer of a print, regardless of
+     * what nozzle is printing.
+     */
+    void babystepAxis_steps(const int16_t steps, const axis_t axis, bool linked_nozzles) {
+      const float mm = steps * planner.steps_to_mm[axis];
+
+      switch (axis) {
+        #if ENABLED(BABYSTEP_XY)
+          case X: thermalManager.babystep_axis(X_AXIS, steps); break;
+          case Y: thermalManager.babystep_axis(Y_AXIS, steps); break;
+        #endif
+        case Z: thermalManager.babystep_axis(Z_AXIS, steps); break;
+        default: return;
+      };
+
+      #if ENABLED(BABYSTEP_ZPROBE_OFFSET)
+        // Make it so babystepping in Z adjusts the Z probe offset.
+        if (axis == Z
+          #if EXTRUDERS > 1
+            && (linked_nozzles || active_extruder == 0)
+          #endif
+        ) zprobe_zoffset += mm;
+      #endif
 
-    void addZOffset_steps(int16_t babystep_increment) {
-      #if ENABLED(BABYSTEP_HOTEND_Z_OFFSET)
-        const bool do_probe = (active_extruder == 0);
+      #if EXTRUDERS > 1
+        /**
+         * When linked_nozzles is false, as an axis is babystepped
+         * adjust the hotend offsets so that the other nozzles are
+         * unaffected by the babystepping of the active nozzle.
+         */
+        if (!linked_nozzles) {
+          HOTEND_LOOP()
+            if (e != active_extruder)
+              hotend_offset[axis][e] += mm;
+
+          #if ENABLED(BABYSTEP_ZPROBE_OFFSET)
+            // Transfer any Z offset in the zeroth hotend to zprobe_zoffset
+            if (axis == Z) {
+              zprobe_zoffset += hotend_offset[Z_AXIS][0];
+              normalizeNozzleOffset(Z);
+            }
+          #endif
```

Comment:
Should this just call `normalizeNozzleOffset(axis)` for all cases, since `[x][0]` or `[y][0]` could (theoretically) be changed by the loop above?

Reply:
The reason I call `normalizeNozzleOffset(Z)` specifically in this line is because I copied `hotend_offset[Z_AXIS][0]` into `zprobe_zoffset`, so I wanted to zero out `hotend_offset[Z_AXIS][0]`. `normalizeNozzleOffset(Z);` does that for me. But as you say, I could normalize all axis if I wanted to, it's not necessary, however.

Truth be told, it's unclear whether even what I am doing here is totally correct. Come to think of it, zprobe_zoffset probably ought to be set the hotend_offset of the *lowest* nozzle.

Reply:
Well, given that `hotend_offset[Z][0]` must definitely be 0, I think what you're doing here makes sense, and the same should also apply for `hotend_offset[X][0]` and `hotend_offset[Y][0]`.

Reply:
> `zprobe_zoffset` probably ought to be set the `hotend_offset` of the lowest nozzle

Why would the `hotend_offset[Z]` have anything to do with the `zprobe_zoffset`? The hotend offset refers to the difference between nozzles where one might be higher or lower than the other. (The value of `hotend_offset[Z]` should always be 0 for setups where the nozzles are statically fixed to the same carriage.)

The `zprobe_zoffset` is specifically the distance between Z at the probe's trigger point and the Z at which the nozzle touches the bed. It should not be affected by operations that apply to `hotend_offset[Z]` unless the first nozzle is selected, and in that case `hotend_offset[Z]` is always zero.

Reply:
@thinkyhead: I believe this code has been revised since, so this may not be relevent now, but originally as the user nudged the nozzles around, it could cause the hotend_offset[Z] for the first nozzle to go negative or positive. This effectively means that the user wanted the nozzle closer or further from the bed. So, in order to maintain the condition that the hotend_offset[Z] for the first nozzle was zero, I would transfer that offset to the zprobe_zoffset.